src/context.rs

```diff
@@ -341,12 +343,115 @@ impl<A: Actor> Context<A> {
         let mut addr_recv = addr_rx.recv_async();
         let mut broadcast_recv = broadcast_rx.recv_async();
 
-        loop {
+        while self.running == RunningState::Running {
+            let (next_msg, unfinished) = {
+                let next_msg = future::select(addr_recv, broadcast_recv);
+                futures_util::pin_mut!(next_msg);
+                match future::select(fut, next_msg).await {
+                    Either::Left((future_res, _)) => return future_res,
+                    Either::Right(tuple) => tuple,
+                }
+            };
+
+            let msg = match next_msg {
+                Either::Left((res, other)) => {
+                    broadcast_recv = other;
+                    addr_recv = addr_rx.recv_async();
+                    Either::Right(res.unwrap())
+                }
+                Either::Right((res, other)) => {
+                    addr_recv = other;
+                    broadcast_recv = broadcast_rx.recv_async();
+                    Either::Left(res.unwrap())
+                }
+            };
+
+            // To avoid broadcast starvation, try receive a broadcast here
+            if let Ok(Some(broadcast)) = broadcast_rx.try_recv() {
+                self.tick(Either::Left(broadcast), actor).await;
+            }
+
+            self.tick(msg, actor).await;
+            fut = unfinished;
+        }
+
+        fut.await
+    }
+
+    /// Handle any incoming messages for the actor while running a given future. This is similar to
+    /// [`Context::join`], but will exit if the actor is stopped, returning the future. Returns
+    /// `Ok` with the result of the future if it was successfully completed, or `Err` with the
+    /// future if the actor was stopped before it could complete.
```

Comment:
Similar to above, I think a comparison to `futures::future::select` would be appropriate here.

Reply:
Good point - I just named them similarly but did not add an explicit comparison